src/arch/generic/semihosting.hh

```diff
@@ -0,0 +1,557 @@
+ * licensed hereunder.  You may use the software subject to the license
+ * terms below provided that you ensure that this notice is replicated
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __ARCH_GENERIC_SEMIHOSTING_HH__
+#define __ARCH_GENERIC_SEMIHOSTING_HH__
+
+#include <cstdio>
+#include <functional>
+#include <map>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/time.hh"
+#include "cpu/thread_context.hh"
+#include "mem/port_proxy.hh"
+#include "sim/core.hh"
+#include "sim/guest_abi.hh"
+#include "sim/sim_object.hh"
+
+namespace gem5
+{
+
+struct BaseSemihostingParams;
+class SerialDevice;
+
+/**
+ * Semihosting for AArch32, AArch64, RISCV-32 and RISCV-64:
+ * https://github.com/ARM-software/abi-aa/blob/main/semihosting/semihosting.rst
+ *
+ * This class implements the Arm semihosting interface. This interface
+ * allows baremetal code access service, such as IO, from the
+ * simulator. It is conceptually a simplified version of gem5's more
+ * general syscall emulation mode.
+ *
+ * Note: While the current (as of Q1 2024) RISC-V semihosting specificion does
+ * not explicitly define the calls that are required to be implemented
+ * (https://github.com/riscv-non-isa/riscv-semihosting), the intention of the
+ * spec is to reuse the Arm interfaces (which is what QEMU also does).
```

Comment:
```suggestion
 * Note: The RISC-V semihosting specificion reuses the Arm semihosting
 * interfaces (https://github.com/riscv-non-isa/riscv-semihosting).
```

Reply:
This has now been fixed in the latest RISC-V semihosting spec.
There is a 0.3 release that says the following:

> The RISC-V semihosting specification adapts the design of publicly available and open source
> semihosting specification [1] to minimize the development effort. Only the mechanism of invoke a
> semihosting service (aka semihosting binary interface) is architecture specific whereas all defined
> semihosting services [1] are architecture independent and resemble system calls of an unix-style
> operating system.

With [1] pointing to the ARM semihosting spec.